zebra-chain/src/notes/sprout.rs

```diff
@@ -1,33 +1,48 @@
 //!
 #![allow(dead_code)]
 
+use super::{memo::Memo, *};
+use crate::serde_helpers;
+use crate::serialization::{SerializationError, ZcashDeserialize, ZcashSerialize};
+use crate::types::amount::{Amount, NonNegative};
 use serde::{Deserialize, Serialize};
+use sha2::{Digest, Sha256};
 use std::{
     fmt,
     io::{self},
 };
 
-#[cfg(test)]
-use proptest::{collection::vec, prelude::*};
-
-use crate::serde_helpers;
-use crate::serialization::{SerializationError, ZcashDeserialize, ZcashSerialize};
-
-use super::{memo::Memo, *};
-
 ///
 pub struct Note {
     // TODO: refine type as a SHA-256d output derived from a spending key.
     paying_key: [u8; 32],
-    value: u64,
+    value: Amount<NonNegative>,
     // TODO: refine type as the input to the PRF that results in a nullifier.
     nullifier_seed: [u8; 32],
     note_commitment_randomness: NoteCommitmentRandomness,
 }
 
+impl Note {
+    pub fn note_commitment(&self) -> NoteCommitment {
+        let leading_byte: u8 = 0xB0;
+        let mut hasher = Sha256::default();
```

Comment:
Pretty sure this is required to be `SHA256Compress` which is a different function. It's the same function as used for deriving Sprout shielded payment addresses from a Sprout spending key:

https://github.com/ZcashFoundation/zebra/blob/7a75522885b2db3c9c868455933d37f5bae348d5/zebra-chain/src/keys/sprout.rs#L47

![image](https://user-images.githubusercontent.com/552961/86857828-58259080-c08d-11ea-81c3-68b749e3cf85.png)

I can fix this as I'm getting in here for the note commitment tree stuff.

Reply:
I'm confused, it doesn't mention compress in the spec https://zips.z.cash/protocol/protocol.pdf#concretesproutnotecommit, also I don't know how I'd apply `compress256` to the data mentioned in the spec which seems to add up to `105` bytes when `compress256` seems to only work with `64` bytes at a time.

Just noticed you said you'd handle this, so I'll stop fiddling with the PR on my end and let you handle it.

Reply:
I just realized I was looking at the nullifier creation in the section labelled ['Note Commitments and Nullfiers'](https://zips.z.cash/protocol/protocol.pdf#commitmentsandnullifiers), so you are correct and I was incorrect. Sorry for the confusion 🙈